### ISSUE TITLE
Group by is not working correctly with all types

### ISSUE THREAD

**Issue by tflori** [laravel/framework#27974]
- Laravel Version: 5.4.36
- PHP Version: 7.1.*
- Database Driver & Version: irrelevant

### Description:
The documentation (https://laravel.com/docs/5.8/collections#method-groupby) does nowhere describe what happens with non integer or string values. Because array_key_exists only allows strings and integers it will fail in all other cases. From code I saw that there is a conversion for booleans to integer. This is also not defined in the documentation.

The more evil thing is that debuggers show floats/doubles without decimals like integers and before you call `gettype($groupKey)` you will never find out that it is a double.

### Steps To Reproduce:
```console
$ artisan tinker
>>> $c = collect([['f' => 1.0, 'i' => 1, 's' => '1', 'dt' => new DateTime()], ['f' => 1.0, 'i' => 1, 's' => '1', 'dt' => new DateTime()]])
>>> $c->groupBy('f')->all()
>>> $c->groupBy('dt')->all()
```

Conclusion: There is a bug in your collection that could be fixed at least but I don't know if you are willing to fix it, in which branch and how you want to solve it. And because I know this community already I'm not willing to create a PR, sorry.

*Hint:* Some cases may be solved by converting items to strings if possible - so `(string)1.0` would still be a valid key for `array_key_exists($key, $array)` and some other types might also work. At least the function should not only check for bool because array_key_exists accepts only string and int.

minimal solution:
```php
$groupKey = is_int($groupKey) ? $groupKey : (string)$groupKey;
```

**Comment by driesvints** [comment]
Hey there,

Unfortunately we don't support this version anymore. Please check out [our support policy](https://laravel.com/docs/releases) on which versions we are currently supporting. Can you please try to upgrade to the latest version and see if your problem persists? If so feel free to reply and we'll try to have a look.

Thanks!

**Comment by tflori** [comment]
See, that's one of the reasons I don't support this communtiy anymore. Instead of looking into the code the issue gets closed.

FYI: the first thing I did was checking the code in all versions up to version 5.8 and the problem is not solved in any version.

**Comment by driesvints** [comment]
> See, that's one of the reasons I don't support this communtiy anymore. Instead of looking into the code the issue gets closed.

Hey @tflori. The reason why I didn't look at your code is because you reported this under version "5.4.36" which we don't support anymore as I explained above. When you open an issue you can see the following link underneath "bug report" which links to a page in the docs with the supported versions:

![Screen Shot 2019-03-25 at 13 12 49](https://user-images.githubusercontent.com/594614/54918956-d0e7f400-4eff-11e9-9599-f37534138bf7.png)

Because we get so much issues a day and I still almost have 300 issues to go through and a lot of other stuff on my plate it's unfeasible for us to provide support for every version out there. That's one of the reasons why we have a support policy with EOL dates for framework versions.

> FYI: the first thing I did was checking the code in all versions up to version 5.8 and the problem is not solved in any version.

Well I said:

> Can you please try to upgrade to the latest version and see if your problem persists? If so feel free to reply and we'll try to have a look.

So I'll have another look.

In this particularly situation I think that maybe a note in the docs would suffice. Feel free to send in a PR if you want.

**Comment by tflori** [comment]
I never said you should look in my code. Your code is throwing unexpected exceptions - that's a problem of your code.

> Group by is not working correctly

That's enough  description to understand that the problem is in the group by method of your code. So you should check your code:
https://github.com/laravel/framework/blob/5.8/src/Illuminate/Support/Collection.php#L854

```php
            foreach ($groupKeys as $groupKey) {
                $groupKey = is_bool($groupKey) ? (int) $groupKey : $groupKey;

                if (! array_key_exists($groupKey, $results)) {
                    $results[$groupKey] = new static;
                }

                $results[$groupKey]->offsetSet($preserveKeys ? $key : null, $value);
            }
```

Before you ignore an issue (because it's an unsupported version the author is using) you should not assume the developer is not using the framework correct or is struggling with a bug that got solved in a later version. And when you read the description you will notice that I already checked *your code* and you would actually save time instead of wasting it for such discussions.

If you use issues for dev support that's your problem not mine. I only create tickets because the code is not correct.

I know you will not fix the issue in version 5.4 - but that does not matter for me (I just extend the collection and overwrite the group by method). I'm not here to give you work for nothing. Instead I'm trying to helping you to find problems in your code. But every time I create an issue here I just get blamed. You should really thing about your communication. There is a thin line between proud and arrogance.
